File: android/src/main/java/com/carlrosales/capacitor/storage/SecurePreferencesPlugin.java
package com.carlrosales.capacitor.storage;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "SecurePreferences")
public class SecurePreferencesPlugin extends Plugin {

    private SecurePreferences implementation;

    @Override
    public void load() {
        implementation = new SecurePreferences(getContext());
    }

    @PluginMethod
    public void set(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");
        if (key == null || value == null) {
            call.reject("key and value are required");
            return;
        }
        implementation.set(key, value);
        call.resolve();
    }

    @PluginMethod
    public void get(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("key is required");
            return;
        }
        String value = implementation.get(key);
        JSObject ret = new JSObject();
        ret.put("value", value);
        call.resolve(ret);
    }

    @PluginMethod
    public void remove(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("key is required");
            return;
        }
        implementation.remove(key);
        call.resolve();
    }

    @PluginMethod
    public void keys(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("keys", implementation.keys());
        call.resolve(ret);
    }

    @PluginMethod
    public void clear(PluginCall call) {
        implementation.clear();
        call.resolve();
    }
}

File: ios/Sources/SecurePreferencesPlugin/SecurePreferencesPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@available(iOS 15.0, *)
@objc(SecurePreferencesPlugin)
public class SecurePreferencesPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SecurePreferencesPlugin"
    public let jsName = "SecurePreferences"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "keys", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = SecurePreferences()

    @objc func set(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), let value = call.getString("value") else {
            call.reject("key and value are required")
            return
        }
        do {
            try implementation.set(key: key, value: value)
            call.resolve()
        } catch {
            call.reject("Failed to set value: \(error.localizedDescription)")
        }
    }

    @objc func get(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("key is required")
            return
        }
        do {
            let value = try implementation.get(key: key)
            call.resolve(["value": value as Any])
        } catch {
            call.reject("Failed to get value: \(error.localizedDescription)")
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("key is required")
            return
        }
        do {
            try implementation.remove(key: key)
            call.resolve()
        } catch {
            call.reject("Failed to remove value: \(error.localizedDescription)")
        }
    }

    @objc func keys(_ call: CAPPluginCall) {
        do {
            let k = try implementation.keys()
            call.resolve(["keys": k])
        } catch {
            call.reject("Failed to get keys: \(error.localizedDescription)")
        }
    }

    @objc func clear(_ call: CAPPluginCall) {
        do {
            try implementation.clear()
            call.resolve()
        } catch {
            call.reject("Failed to clear: \(error.localizedDescription)")
        }
    }
}
